keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -40,74 +56,145 @@ For operators who want to regularly rotate the signing and verifying keys for pr
+
+### New API
+
+I'm proposing creating a new versioned grpc API (KeyService) under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`. It will have methods to Sign and JWT and Fetch supported keys (proto in Details section). 
+
+### Plugins
+
+- A signer plugin will be a client to the new grpc API thus providing support for external Signing.
+- A public key Cache will integrate with the same grpc to fetch supported public keys.
 
 ### Updates to API server token generation
 
-As of Kubernetes v1.13.2, the API server uses the functions `JWTTokenGenerator` and `JWTTokenAuthenticator`. New types that implement the `TokenGenerator` interface and support token validation will be added to `k8s.io/kubernetes/pkg/serviceaccount/`.
+As of Kubernetes `v1.13.2`, The Kubernetes Control Plane config accepts `ServiceAccountIssuer` which is an interface of type `serviceaccount.TokenGenerator` as defined in `pkg/serviceaccount/jwt.go`. This enables defining alternate implementations of token generators. Thus, the default TokenGenerator can be replaced by one that integrates with KeyService.
 
-### New API
+### Updates for supported public keys
 
-I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+Starting ` v1.31` (via. the [PR](https://github.com/kubernetes/kubernetes/pull/125177) by liggit@), `ServiceAccountPublicKeys` in ApiServer config was replaced by `ServiceAccountPublicKeysGetter` interface. The static keys held by `ServiceAccountPublicKeys` were hence moved to be managed by the implementation called `StaticPublicKeysGetter`. This presents an opportunity to add a new implementation, one that will get public keys from KeyService depending on how the cluster is configured.
 
-```proto
-syntax = "proto3";
+In the same PR, `OpenIDMetadataServer` was updated to accept implementation of `OpenIDMetadataProvider` as defined in  `pkg/serviceaccount/openidmetadata.go`. `OpenIDMetadataProvider` interacts with `ServiceAccountPublicKeysGetter` for serving supported keys which enables flexible replacing `StaticPublicKeysGetter` with a dynamic key getter (will integrate with KeyService) depending on how cluster is configured.
+
+### Notes/Constraints/Caveats
+
+- A new flag `--key-service-url` will be added to the API server specifying a unix socket where the key service will be accessible.
+- cumulation on local and external keys will not be supported.
+- If Both, `--service-account-key-file` and `--key-service-url` are specified, keys from external signers will override the keys at `service-account-key-file` as supported public keys.
```

Comment:
We should probably just throw an error if this happens?

Also, "Both" should not be capitalized.

Reply:
done and done